exp/controllerprincipalcreator/awscontrollerprincipal_controller.go

```diff
@@ -0,0 +1,161 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllerprincipalcreator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha3"
+	"sigs.k8s.io/cluster-api-provider-aws/controllers"
+	controlplanev1 "sigs.k8s.io/cluster-api-provider-aws/controlplane/eks/api/v1alpha3"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/scope"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// AWSControllerPrincipalReconciler reconciles a AWSClusterControllerPrincipal object
+type AWSControllerPrincipalReconciler struct {
+	client.Client
+	Log       logr.Logger
+	Endpoints []scope.ServiceEndpoint
+}
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=awsclustercontrollerprincipals,verbs=get;list;watch;create
+
+func (r *AWSControllerPrincipalReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.TODO()
+	log := r.Log.WithValues("namespace", req.NamespacedName, "awsControllerPrincipal", req.Name)
+
+	var principalRef *v1.ObjectReference
+
+	// Fetch the AWSCluster instance
+	awsCluster := &infrav1.AWSCluster{}
+	clusterFound := true
+	err := r.Get(ctx, req.NamespacedName, awsCluster)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+		log.V(2).Info("AWSCluster not found, trying AWSManagedControlPlane")
+		clusterFound = false
+	} else {
+		log.V(2).Info("Found principleRef on AWSCluster")
+		principalRef = awsCluster.Spec.PrincipalRef
+	}
+
+	// If AWSCluster is not found, check if AWSManagedControlPlane is used.
+	if !clusterFound {
+		awsControlPlane := &controlplanev1.AWSManagedControlPlane{}
+		if err := r.Client.Get(ctx, req.NamespacedName, awsControlPlane); err != nil {
+			if apierrors.IsNotFound(err) {
+				log.V(2).Info("AWSManagedMachinePool not found, no principleRef so no action taken")
+				return ctrl.Result{}, nil
+			}
+			return reconcile.Result{}, err
+		}
+		log.V(2).Info("Found principleRef on AWSManagedControlPlane")
+		principalRef = awsControlPlane.Spec.PrincipalRef
+	}
+
+	if principalRef == nil {
+		log.Info("PrincipalRef is nil, skipping reconciliation", "cluster", req.Name)
+		return ctrl.Result{Requeue: true}, nil
```

Comment:
@sedefsavas @randomvariable - i added this test for a nil principleRef. Currently, it requeues but should it error?

Reply:
There is a defaulting in both AWSCluster and AWSManagedControlPlane that sets principleRef to `AWSClusterControllerPrincipal` if it is nil.

Reply:
Ok cool, this can been removed 😁(I did encounter this being nil in my testing but that was probably because of the AWSCluster logic that had now be changed above)